File: solutions/ruby/tournament/1/tournament.rb
class Tournament
  def self.tally(input)
    scores = Hash.new { |h, k| h[k] = { w: 0, d: 0, l: 0 } }

    input.strip.split("\n").each do |line|
      team1, team2, result = line.split(';')
      case result
      when 'win'
        scores[team1][:w] += 1
        scores[team2][:l] += 1
      when 'loss'
        scores[team1][:l] += 1
        scores[team2][:w] += 1
      when 'draw'
        scores[team1][:d] += 1
        scores[team2][:d] += 1
      end
    end

    sorted_scores = scores.map do |team, s|
      mp = s[:w] + s[:d] + s[:l]
      p = s[:w] * 3 + s[:d]
      { team: team, mp: mp, w: s[:w], d: s[:d], l: s[:l], p: p }
    end.sort_by { |s| [-s[:p], s[:team]] }

    output = "Team                           | MP |  W |  D |  L |  P\n"
    sorted_scores.each do |s|
      output += sprintf("%-30s | %2d | %2d | %2d | %2d | %2d\n", s[:team], s[:mp], s[:w], s[:d], s[:l], s[:p])
    end
    output
  end
end

File: solutions/python/tournament/1/tournament.py
def tally(rows):
    stats = {}
    for row in rows:
        team1, team2, result = row.split(';')
        if team1 not in stats:
            stats[team1] = {'mp': 0, 'w': 0, 'd': 0, 'l': 0, 'p': 0}
        if team2 not in stats:
            stats[team2] = {'mp': 0, 'w': 0, 'd': 0, 'l': 0, 'p': 0}
        if result == 'win':
            stats[team1]['w'] += 1
            stats[team1]['p'] += 3
            stats[team2]['l'] += 1
        elif result == 'loss':
            stats[team2]['w'] += 1
            stats[team2]['p'] += 3
            stats[team1]['l'] += 1
        elif result == 'draw':
            stats[team1]['d'] += 1
            stats[team1]['p'] += 1
            stats[team2]['d'] += 1
            stats[team2]['p'] += 1
        stats[team1]['mp'] += 1
        stats[team2]['mp'] += 1
    teams = sorted(stats.items(), key=lambda x: (-x[1]['p'], x[0]))
    header = f"{'Team':31s}| MP |  W |  D |  L |  P"
    table = [header]
    for name, s in teams:
        table.append(
            f"{name:31s}| {s['mp']:>2d} | {s['w']:>2d} | {s['d']:>2d} | {s['l']:>2d} | {s['p']:>2d}"
        )
    return table
